File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        r, c = len(matrix), len(matrix[0])
        t, b = 0, len(matrix) - 1

        while t <= b:
            row = (t + b) // 2
            if target > matrix[row][-1]:
                t = row + 1
            elif target < matrix[row][0]:
                b = row - 1
            else:
                break
        
        if not (t <= b):
            return False
        
        row = (t + b) // 2
        l, r = 0, len(matrix[row]) - 1
        while l <=r:
            m = l + (r - l) // 2
            if target < matrix[row][m]:
                r = m - 1
            elif target > matrix[row][m]:
                l = m + 1
            elif target == matrix[row][m]:
                return True
        return False
                

File: Data Structures & Algorithms/two-integer-sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums, target) {
        let map = new Map();
        for (let i = 0; i < nums.length; i++) {
            map[nums[i]] = i
        }

        for (let i = 0; i < nums.length; i++) {
            let diff = target - nums[i]
            let j = map[diff]
            if (j) {
                if (i == j) continue;
                if (i < j) {
                    return [i, j]
                } else {
                    return [j, i]
                }
            }
        }

        return [0,0]
    }
}
